### ISSUE TITLE
Failing tests due to DUPLICATE_USERNAME

### ISSUE THREAD

**Issue by jverelst** [jongpie/NebulaLogger#313]
### Package Edition of Nebula Logger

Unlocked Package

### Package Version of Nebula Logger

v4.6.10 - current

### New Bug Summary

We've been seeing some errors when trying to do a full test run recently, which seem to be avoidable:

```
LogBatchPurger_Tests.it_should_throw_exception_when_user_does_not_have_delete_access
System.DmlException: Insert failed. First exception on row 0; first error: DUPLICATE_USERNAME, Duplicate Username.<br>The username already exists in this or another Salesforce organization. Usernames must be unique across all Salesforce organizations. To resolve, use a different username (it doesn't need to match the user's email address). : [Username]  

Class.LogBatchPurger_Tests.it_should_throw_exception_when_user_does_not_have_delete_access: line 46, column 1                                                                                                                                                                                                                                                                   
System.DmlException: Insert failed. First exception on row 0; first error: DUPLICATE_USERNAME, Duplicate Username.<br>The username already exists in this or another Salesforce organization. Usernames must be unique across all Salesforce organizations. To resolve, use a different username (it doesn't need to match the user's email address). : [Username]  

Class.LogEntryEventBuilder_Tests.it_should_skip_stripping_inaccessible_fields_for_aggregate_result: line 808, column 1                                                                                                                                                                                                                                                          
System.DmlException: Insert failed. First exception on row 0; first error: DUPLICATE_USERNAME, Duplicate Username.<br>The username already exists in this or another Salesforce organization. Usernames must be unique across all Salesforce organizations. To resolve, use a different username (it doesn't need to match the user's email address). : [Username]  
                                                                                                                               Class.LogEntryEventBuilder_Tests.it_should_skip_stripping_inaccessible_fields_for_aggregate_results: line 822, column 1                                                                                                                                                                                                                                                         
```

There seems to be an issue in the way the test user is being generated: the username is not made fully unique, so it can clash with other usernames.

In our version v4.6.10 the problem is located in the `LoggerTestUtils.createStandardUser()` method. I saw that in the current codebase, it's still not entirely fixed in `LoggerMockDataCreator.createUser()`:

```
    /**
     * @description Creates a `User` record for testing purposes, using the specified profile ID
     * @param  profileId The `Profile` ID to use for the created `User`
     * @return   The generated `User` record - it is not automatically inserted into the database.
     */
    public static User createUser(Id profileId) {
        return new User(
            Alias = 'user_xyz',
            Email = 'user_xyz@test.com.net.org',
            EmailEncodingKey = 'ISO-8859-1',
            LanguageLocaleKey = 'en_US',
            LastName = 'Test Userson',
            LocaleSidKey = 'en_US',
            ProfileId = profileId,
            TimeZoneSidKey = 'America/Los_Angeles',
            Username = 'user_xyz' + (userMockUsernameCount++) + '@test.com.net.org'
        );
    }
```

If I were to create a user called 'user_xyz1@est.com.net.org' in my org, tests would fail in other orgs I assume. 

I think it would be most robust if there was a org-specific part (for instance the orgId, a timestamp, etc.) in the username, so it can never clash.

**Comment by jongpie** [comment]
Hi @jverelst - I think the reason you have these errors is because in `v4.6.10`, [the same hardcoded username was used for any test that created a standard users](https://github.com/jongpie/NebulaLogger/blob/6637526a18f4975a97e7564f27385d61c5fad91a/nebula-logger/tests/common/classes/LoggerTestUtils.cls#L50), which could result in the `DUPLICATE_USERNAME` exception. This is was fixed as part of [the `v4.7.1` release](https://github.com/jongpie/NebulaLogger/releases/v4.7.1) by [incrementing a number within the username each time the method is called](https://github.com/jongpie/NebulaLogger/blob/3b04138cdb22211796870ef49d8b3773465641e9/nebula-logger/core/tests/configuration/utilities/LoggerMockDataCreator.cls#L314), which should avoid the issue when running tests in parallel.

I don't expect there to be any more `DUPLICATE_USERNAME` exceptions in v4.7.1 or newer (mostly because it seems unlikely someone would actually create a user with the username 'user_xyz1@est.com.net.org'), but I can still include the org ID as part of the username to be safe - that seems like the simplest option to try to ensure uniqueness. I'll include that change in the next release, `v4.7.3`

**Comment by jongpie** [comment]
@jverelst I've just released [v4.7.3](https://github.com/jongpie/NebulaLogger/releases/tag/v4.7.3), which includes a change to how usernames are generated in the class `LoggerMockDataCreator` - it now uses `Logger.getTransactionId()` to create a truly unique username across orgs/transactions (`getTransactionId()` in v4.7.3 now always generates a UUID), and an incremented counter to ensure that every username is also unique within each transaction. 

Thanks again for your feedback - let me know if you have any other suggestions!
